### PR TITLE
Thread safe caching of module info in HLTConfigData

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Excessive numbers of allocations per event were seen from users of this code.

#### PR validation:

Using step3 of workflow 12861.0 and the ModuleAllocMonitor the number of allocations per event measured from a module using this code went from 155k to 37k. This was tested using 8 threads.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
@makortel FYI

**Comment by Dr15Jones** [comment]
please test

**Comment by Martin-Grunewald** [comment]
What is ```#define CMS_SA_ALLOW [[cms::sa_allow]]``` ?

**Comment by fwyzard** [comment]
It's an attribute used to tell the static analyser not to complain about a piece of code.

**Comment by Martin-Grunewald** [comment]
So this is to avoid ```exists``` or ```existsAs`` in normal code?

**Comment by Dr15Jones** [comment]
> So this is to avoid exists or existsAs in normal code?

We have told the static analyzer to give warnings when those two are used in code as they have proven to be problematic. The attribute turns off that warning.

**Comment by Dr15Jones** [comment]
I have some minor updates related to this PR (e.g. changing HLTConfigProvider to also return `const std::string&`). These further changes decrease the allocations by a further 8%. Should I add them to this PR or wait to make an additional one?

**Comment by Martin-Grunewald** [comment]
Thanks @fwyzard and @Dr15Jones 
For me it would be fine to add the minor updates to this PR...

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
type performance-improvements

**Comment by makortel** [comment]
In step3 of 2024.202001, the number of allocations decreases from [2372298522](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_14_2_X_2024-11-07-1100/el8_amd64_gcc12/-GenuineIntel/matrix-results/2024.202001_RunJetMET02024D_10k/step3_RunJetMET02024D_10k.log) to [2308247017](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5cb79f/42682/runTheMatrix-results/2024.202001_RunJetMET02024D_10k/step3_RunJetMET02024D_10k.log), i.e. by 64 million (2.7 %, for 100 events).

File: HLTrigger/HLTcore/interface/HLTConfigData.h [hunk 1]
```diff
@@ -151,6 +152,10 @@ class HLTConfigData {
   edm::ParameterSetID id() const;
 
 private:
+  inline std::string canonicalModuleName(const std::string& module) const {
+    return module.front() != '-' ? module : module.substr(1);
+  }
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply HLTrigger/HLTcore/interface/HLTConfigData.h:157]: I actually tried that at one point but it didn't work with `tbb::concurrent_unordered_map` so didn't seem worth the change.

**Comment by Martin-Grunewald** [comment]
+1

**Comment by jfernan2** [comment]
+1

**Comment by hqucms** [comment]
+1

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**